components/script/dom/element.rs

```diff
@@ -864,29 +864,21 @@ impl Element {
         inline: ScrollLogicalPosition,
         container: Option<&Element>,
     ) {
-        // Shadow-inclusive ancestors of this node, skipping the node itself.
-        //
-        // TODO: This should be the ancestors of the node in the flat tree.
-        let ancestors = self
-            .upcast::<Node>()
-            .inclusive_ancestors(ShadowIncluding::Yes)
-            .skip(1);
-
         // Step 1: For each ancestor element or viewport that establishes a scrolling box `scrolling
         // box`, in order of innermost to outermost scrolling box, run these substeps:
-        for node in ancestors {
-            if !node.establishes_scrolling_box() {
-                continue;
-            }
+        let mut maybe_scroll_parent = self.owner_window().scroll_parent_query(self.upcast());
+        while let Some(scroll_parent) = maybe_scroll_parent {
+            maybe_scroll_parent = scroll_parent
+                .owner_window()
+                .scroll_parent_query(scroll_parent.upcast());
 
             // TODO: This should skip elements which are not containing blocks of the element we
             // are scrolling.
-            let scrolling_box = if let Some(document) = node.downcast::<Document>() {
-                ScrollingBox::Viewport(DomRoot::from_ref(document))
-            } else if let Some(element) = node.downcast::<Element>() {
-                ScrollingBox::Element(DomRoot::from_ref(element))
+            let document = scroll_parent.owner_document();
+            let scrolling_box = if Some(&scroll_parent) == document.GetScrollingElement().as_ref() {
```

Comment:
Instead of using `scroll_parent_query()` and then comparing with `document.GetScrollingElement()`, I think this should call `query_scroll_parent` directly, mapping `ScrollParentResponse::DocumentScrollingElement` to `ScrollingBox::Viewport(document)`.

Also, I think we need to add another response for the root and the body. `scrollParent` returns null to avoid `document.documentElement.scrollParent === document.documentElement`, but they may still be moved by scrolling the viewport.